### PR TITLE
Clear storage during genesis block initialization

### PR DESCRIPTION
**Pull request by kamilsa**
## What is this pull request?
1. Add clearing storage before genesis block insertion
2. Allow running iroha if it was already initialized using genesis block
   
## Why do you implement it? Why do we need this pull request?
Lack of clearing storage before genesis block insertion leads to error when irohad has been launched before. 
  
## How to use the features provided in the pull request?
To overcome the issue now storage is cleared out when genesis block flag was passed into irohad. If not irohad launches from state it achieved during previous execution.

## Details/Features
To drop storage irohad class was updated with method dropStorage, which uses TestStorage to drop existing storage.

### CODE REVIEW

File: irohad/main/application.cpp [hunk 1]
```diff
@@ -41,7 +41,9 @@ Irohad::Irohad(const std::string &block_store_dir,
       pg_conn_(pg_conn),
       torii_port_(torii_port),
       internal_port_(internal_port),
-      keypair(keypair) {
+      keypair(keypair),
+      test_storage_(TestStorageImpl::create(
```
Comment (grimadas) [inline irohad/main/application.cpp]: Why using test storage ?

Comment (kamilsa) [reply irohad/main/application.cpp]: Storage cannot drop database while `TestStorage` can. Probably it should be renamed to `DroppableStorage`

File: irohad/main/application.hpp [hunk 2]
```diff
@@ -167,6 +173,9 @@ class Irohad {
   iroha::consensus::yac::YacInit yac_init;
   iroha::network::BlockLoaderInit loader_init;
 
+  // drop storage
```
Comment (muratovv) [inline irohad/main/application.hpp]: This comment is very confused

File: irohad/main/irohad.cpp [hunk 3]
```diff
@@ -17,6 +17,8 @@
 
 #include <gflags/gflags.h>
 #include <grpc++/grpc++.h>
+#include <ametsuchi/impl/test_storage_impl.hpp>
```
Comment (muratovv) [inline irohad/main/irohad.cpp]: Use quotes instead.

File: irohad/main/irohad.cpp [hunk 4]
```diff
@@ -80,20 +78,25 @@ int main(int argc, char *argv[]) {
     return EXIT_FAILURE;
   }
 
-  iroha::main::BlockInserter inserter(irohad.storage);
-  auto file = inserter.loadFile(FLAGS_genesis_block);
-  auto block = inserter.parseBlock(file.value());
-  log->info("Block is parsed");
+  if (not FLAGS_genesis_block.empty()) {
+    // clear previous storage if any
+    irohad.dropStorage();
```
Comment (muratovv) [inline irohad/main/irohad.cpp]: Move drop storage below. After checking genesis block.

File: irohad/main/application.cpp [hunk 5]
```diff
@@ -41,7 +41,9 @@ Irohad::Irohad(const std::string &block_store_dir,
       pg_conn_(pg_conn),
       torii_port_(torii_port),
       internal_port_(internal_port),
-      keypair(keypair) {
+      keypair(keypair),
```
[the remark this answers is not in the document]
Comment (kamilsa) [reply irohad/main/application.cpp]: Ask the one who created it

File: irohad/ametsuchi/impl/storage_impl.cpp [hunk 6]
```diff
@@ -118,6 +118,59 @@ namespace iroha {
           std::move(command_executors.value()));
     }
 
+    bool StorageImpl::insertBlock(model::Block block) {
+      log_->info("create mutable storage");
+      auto storage = createMutableStorage();
+      auto inserted = storage->apply(
+          block,
+          [](const auto &current_block, auto &query, const auto &top_hash) {
```
[the remark this answers is not in the document]
Comment (kamilsa) [reply irohad/ametsuchi/impl/storage_impl.cpp]: In my opinion readability is better when param names exist

File: irohad/ametsuchi/impl/storage_impl.cpp [hunk 7]
```diff
@@ -118,6 +118,59 @@ namespace iroha {
...
+)";
+
+      // erase db
+      log_->info("drop dp");
+      pqxx::connection connection(postgres_options_);
+      pqxx::work txn(connection);
+      txn.exec(drop);
+      txn.commit();
+
+      pqxx::connection init_connection(postgres_options_);
```
[the remark this answers is not in the document]
Comment (kamilsa) [reply irohad/ametsuchi/impl/storage_impl.cpp]: thanks, it seems to be unused. I will remove it

File: irohad/ametsuchi/impl/storage_impl.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (l4l) [reply irohad/ametsuchi/impl/storage_impl.cpp]: what's the matter if you don't use them? compiler don't help you with mismatching names anyway

**Pull request merged**